frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -211,219 +230,290 @@ describe("QueryBuilder", () => {
+        const { mockEventListener } = await setup({
+          card: TEST_UNSAVED_STRUCTURED_CARD,
         });
 
-        userEvent.tab();
+        expect(screen.queryByText("Count")).not.toBeInTheDocument();
+        userEvent.click(await screen.findByText("Summarize"));
+        userEvent.click(await screen.findByText("Done"));
+        expect(await screen.findByText("Count")).toBeInTheDocument();
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
 
-      it("should not trigger beforeunload event when model metadata is unedited", async () => {
+      it("should not trigger beforeunload event when query is unedited", async () => {
         const { mockEventListener } = await setup({
-          card: TEST_MODEL_CARD,
-          dataset: TEST_MODEL_DATASET,
-          initialRoute: `/model/${TEST_MODEL_CARD.id}/metadata`,
+          card: TEST_STRUCTURED_CARD,
         });
 
         const mockEvent = callMockEvent(mockEventListener, "beforeunload");
         expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-        expect(mockEvent.returnValue).toBe(undefined);
+        expect(mockEvent.returnValue).not.toEqual(
+          BEFORE_UNLOAD_UNSAVED_MESSAGE,
+        );
       });
     });
   });
 
-  describe("beforeunload events in native queries", () => {
-    afterEach(() => {
-      jest.restoreAllMocks();
-    });
-
-    it("should trigger beforeunload event when leaving edited question", async () => {
-      const { mockEventListener } = await setup({
+  describe("downloading results", () => {
+    it("should allow downloading results for a native query", async () => {
+      const mockDownloadEndpoint = fetchMock.post(
+        `/api/card/${TEST_NATIVE_CARD.id}/query/csv`,
+        {},
+      );
+      await setup({
         card: TEST_NATIVE_CARD,
+        dataset: TEST_NATIVE_CARD_DATASET,
       });
 
       const inputArea = within(
         screen.getByTestId("mock-native-query-editor"),
       ).getByRole("textbox");
 
-      userEvent.click(inputArea);
-      userEvent.type(inputArea, "0");
-
-      userEvent.tab();
+      expect(inputArea).toHaveValue("SELECT 1");
 
-      // default native query is `SELECT 1`
-      expect(inputArea).toHaveValue("SELECT 10");
+      userEvent.click(screen.getByTestId("download-button"));
+      userEvent.click(await screen.findByRole("button", { name: ".csv" }));
 
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).toHaveBeenCalled();
-      expect(mockEvent.returnValue).toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
+      expect(mockDownloadEndpoint.called()).toBe(true);
```

Comment:
Considering the limitations you mentioned [here](https://github.com/metabase/metabase/pull/32251/files?diff=split&w=1#r1258064446), maybe you should leave a comment mentioning that the actual CSV download & verification is in that file? I think this test is valuable but for someone with no context of the other test, they might be wondering why you just checked if the function is called rather than its behavior

Reply:
Good idea yeah.

Reply:
Fixed in https://github.com/metabase/metabase/pull/32251/commits/671a99172141a01f2781414a644883de1fae32dd